File: CombinationSum.cpp

/**
 * Basic DSA - Combination Sum (LeetCode 39)
 * Find all unique combinations that sum to target (reuse allowed).
 */
#include <iostream>
#include <vector>

void backtrack(std::vector<std::vector<int>>& ans, std::vector<int>& cur,
               std::vector<int>& cand, int target, int start) {
    if (target == 0) { ans.push_back(cur); return; }
    if (target < 0) return;
    for (int i = start; i < (int)cand.size(); i++) {
        cur.push_back(cand[i]);
        backtrack(ans, cur, cand, target - cand[i], i);
        cur.pop_back();
    }
}

std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates, int target) {
    std::vector<std::vector<int>> ans;
    std::vector<int> cur;
    backtrack(ans, cur, candidates, target, 0);
    return ans;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Combination Sum\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> c1 = {2, 3, 6, 7};
    auto r1 = combinationSum(c1, 7);
    std::cout << "[2,3,6,7], target=7 -> ";
    for (auto& v : r1) {
        std::cout << "[";
        for (size_t i = 0; i < v.size(); i++) std::cout << (i ? "," : "") << v[i];
        std::cout << "] ";
    }
    std::cout << "\n";

    std::vector<int> c2 = {2, 3, 5};
    auto r2 = combinationSum(c2, 8);
    std::cout << "[2,3,5], target=8 -> ";
    for (auto& v : r2) {
        std::cout << "[";
        for (size_t i = 0; i < v.size(); i++) std::cout << (i ? "," : "") << v[i];
        std::cout << "] ";
    }
    std::cout << "\n";
}

File: CombinationSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Basic DSA Problems - One Fundamental Backtracking Question
 * Simple and essential DSA problem for beginners
 */
public class CombinationSum {

    // Problem 1: Combination Sum
    // Distinct positive candidates; reuse unlimited times; combinations that sum to target.

    /**
     * Sort candidates; DFS with start index so order is non-decreasing in combination
     * Time Complexity: roughly O(2^(target/min)) in worst case
     * Space Complexity: O(target/min) recursion depth
     */
    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        if (candidates == null) {
            return null;
        }

        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();
        backtrack(candidates, target, 0, new ArrayList<>(), result);
        return result;
    }

    private static void backtrack(
            int[] candidates,
            int remain,
            int start,
            List<Integer> path,
            List<List<Integer>> result) {
        if (remain == 0) {
            result.add(new ArrayList<>(path));
            return;
        }
        for (int i = start; i < candidates.length && candidates[i] <= remain; i++) {
            path.add(candidates[i]);
            backtrack(candidates, remain - candidates[i], i, path, result);
            path.remove(path.size() - 1);
        }
    }

    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Combination Sum");
        System.out.println("=".repeat(60));

        int[] a = {2, 3, 6, 7};
        System.out.println("target 7 -> " + combinationSum(a, 7));

        int[] b = {2, 3, 5};
        System.out.println("target 8 -> " + combinationSum(b, 8));

        System.out.println("null -> " + combinationSum(null, 10));
    }
}